### ISSUE TITLE
useWebSocket in a for loop

### ISSUE THREAD

**Issue by YikSanChan** [robtaussig/react-use-websocket#40]
Hi,

I have a use case that a react component has to subscribe to multiple topics for real-time rendering. Let's say there is a web page that shows account balance for multiple bank accounts. Each update operation (for example PUT /account/:account_id/balance) will trigger a message sent to the react component that is using socket URL `ws://localhost:8080/ws/account/1`. This way, the react component wants to subscribe to all accounts the user has. However, `useWebSocket` or in general any react hook is not allowed to stay inside a for loop like below:

```javascript
accountIds.forEach(accountId => {
  const topic = `account:${accountId}`;
  const [sendMessage, lastMessage, readyState, getWebSocket] = useWebSocket(
    `${SOCKET_URL}/${topic}`
  );
  ...
})
```

A walkaround I can think of: have the whole component subscribes to `ws://localhost:8080/ws/accounts`, and on message received, check the account id. With this approach, the account id is sent in the payload rather than encoded as part of the socket URL.

What're your suggestions on the use case? Am I misusing your hook? Thanks

**Comment by dimaqq** [comment]
Is it about multiple websocket connections?
In that case I'd suggest wrapping each into a component and rendering multiple components.

Or is about a single connection with multiple outgoing messages?
In that case I don't know...

**Comment by robtaussig** [comment]
@YikSanChan,

I don't think you are misusing my hook, exactly, but you do need to follow React hook rules in the sense that you cannot call hooks inside of a loop. As @dimaqq, suggested, your pattern would probably be most closely replaced by rendering a component for each account, and having each component use `useWebSocket`. That said, I have a few thoughts/questions:

1) What is the upper bound of accounts a user might have? I'm not sure about Safari/IE/Edge, but Chrome and Firefox allow for a maximum of 255 and 200 simultaneous connections. And while this might be far above what you'd expect a single user to need, an admin view might need to track all accounts, in which case 255 and 200 max connections is too low. Max limit aside, each connection imposes a cost to both client and server that could be avoided if you can achieve the same goal with a single connection, which leads me to --

2) Your second idea of subscribing to an `/accounts` endpoint sounds good. Though I would probably take it a step further and have an endpoint like: `ws/user/:userId/accounts, which would receive updates to all accounts by that user -- `ws/accounts`, to me, would suggest that it triggers messages for updates to all accounts regardless of user -- and even if you filter out messages client-side that affect an account not owned by that user, the messages are still hitting the browser and should not be exposed to everyone.

3) Another possibility is using a shared websocket by multiple components (each representing an account). Let's say you have a list of accounts by user:

```tsx

// accountIds = [1,2,3,4,5];
const Accounts = ({ userId, accountIds }) => {
  return (
    <ul>
      {accountIds.map(accountId => {
        return (
          <Account
            key={`account-${accountId}`} //Assumes unique account ids
            userId={userId}
            id={accountId}
          />
        );
      })}
    </ul>
  );
}

const updateAccountInfo = (accountInfo, update) => {
  const nextAccountInfo = { ...accountInfo };

  //Use update data to modify nextAccountInfo;

  return nextAccountInfo;
};

const Account = ({ userId, id }) => {
    const [accountInfo, setAccountInfo] = useState({ accountBalance: 0 });
    const STATIC_OPTIONS = useMemo(() => ({
      share: true,
      filter: message => message.data.accountId === id,
    }), [id]); //id will never change since it is used as part of the key -- if it changes, a new Account component is mounted
  const [sendMessage, lastMessage, readyState, getWebSocket] = useWebSocket(`${SOCKET_URL}/user/${userId}/accounts`);

  useEffect(() => {
    if (lastMessage !== null) {
      setAccountInfo(prev => updateAccountInfo(prev, lastMessage.data));
    }
  }, [lastMessage, sendMessage, setAccountInfo]); //sendMessage and setAccountInfo are stable, and so this effect is only called when lastMessage changes

  return (
    <AccountInfo
      info={accountInfo}
    />
  );
};

const AccountInfo = ({ info }) => {
  return (
    <span>
      Account balance is ${info.accountBalance}
    </span>
  );
};
```

With the above code, you only create one websocket, but each account component maintains a semblance of autonomy as they only update when an incoming message affects the account matching their accountId.

**Comment by robtaussig** [comment]
Not to get too off-topic, but in the above example, you might also want to hoist the accountInfo into a reducer managed by the root component, so as to allow functionality that requires information from all accounts (like a summary, etc).
